src/v/cluster/scheduling/types.cc

```diff
@@ -91,13 +91,74 @@ allocated_partition::allocated_partition(
   , _domain(domain)
   , _state(state.weak_from_this()) {}
 
-void allocated_partition::add_replica(model::broker_shard replica) {
+std::optional<allocated_partition::previous_replica>
+allocated_partition::prepare_move(model::node_id prev_node) {
+    previous_replica prev;
+    auto it = std::find_if(
+      _replicas.begin(), _replicas.end(), [prev_node](const auto& bs) {
+          return bs.node_id == prev_node;
+      });
+    if (it == _replicas.end()) {
+        return std::nullopt;
+    }
+    prev.bs = *it;
+    prev.idx = it - _replicas.begin();
+    if (!_original) {
+        _original.emplace(_replicas.begin(), _replicas.end());
+    } else if (!_original->contains(prev.bs)) {
+        // if replica prev.bs is not original, pick an arbitrary original one
+        // that is not present in the current set as the "source" for prev.bs
+        // (they are all interchangeable).
```

Comment:
nit: had to stare at this for a while to understand.. I think a simple example with 3 replicas would help a reader.

Reply:
makes sense, I'll add examples